StyraInc/opa-csharp
# docs: Refactors + DocFX auto-docs system

## What changed
This PR wraps up some refactoring/bugfix changes that haven't landed on `main` yet, along with the basic plumbing for [DocFX](https://dotnet.github.io/docfx/index.html) automatic reference docs generation.

TODOs:
 - [x] High-Level API usage examples in `README.md`. (See the [Typescript SDK README](https://github.com/StyraInc/opa-typescript/blob/main/README.md#sdk-example-usage-high-level) for examples of the format)
 - [x] Github Actions publishing workflow. ([Ref](https://dotnet.github.io/docfx/index.html#publish-to-github-pages))
   - [x] `sed` scripting to fix the (now) broken Speakeasy Markdown docs URLs that no longer point to anything. (These links now *should* point to the generated docs from DocFX.)
   - Patching the generated README after Speakeasy generation runs can be the subject of a future PR. :sweat_smile: 

## How to test locally

 - Git clone this branch.
 - `dotnet tool update -g docfx` (Install the `docfx` tool)
 - `docfx docs/docfx.json --serve`
 - Visit the generated localhost URL to see the docs!

You should expect to see a landing page like the following:

<details><summary><b>Screenshot</b></summary>

![image](https://github.com/StyraInc/opa-csharp/assets/1906841/a0367747-b1fa-43c1-9774-39aa9dbdd12a)

</details>

If you visit the "API Reference" tab, you should expect to see something like this:

<details><summary><b>Screenshot</b></summary>

![image](https://github.com/StyraInc/opa-csharp/assets/1906841/191e1afd-c8b3-4811-aa9a-844d90f07eab)

</details>
